Added LOCAL time scale

vermashresth opened the pull request:
issue #6487 
Added the ability to use `local` time scale for `Time` and `TimeDelta`.  Operations that are valid now

Subtracting two `Time` instances of 'local' or `None` type.
Subtracting/Adding `Time` and `TimeDelta` instances of `local` or `None` type.
`local` type  `Time` or `TimeDelta` can only be converted to/from `local` type, all other types are invalid.

Also made changes to existing tests, because not all transformations between `Time` are valid now.

mhvk commented on astropy/time/core.py:
```diff
@@ -30,11 +30,14 @@
 
 
-__all__ = ['Time', 'TimeDelta', 'TIME_SCALES', 'TIME_DELTA_SCALES',
+__all__ = ['Time', 'TimeDelta', 'TIME_SCALES','TIME_TYPES', 'TIME_DELTA_SCALES',
```
No need to export `TIME_TYPES`

mhvk commented on astropy/time/core.py:
```diff
@@ -1336,14 +1340,21 @@ def __sub__(self, other):
         else:  # T - T
+
+            # the scales should be compatible (e.g., cannot convert TDB to LOCAL)
+            if(self.scale is not None and self.scale not in other.SCALES or
```
For `Time`, it should not be possible for `self.scale` or `other.scale` to be `None`. Also, the sets are symmetric, so one only needs to check that `other.scale in self.SCALES`.

mhvk commented on astropy/time/tests/test_basic.py:
```diff
@@ -11,7 +11,7 @@
 from ...tests.helper import catch_warnings
 from ...utils import isiterable
-from .. import Time, ScaleValueError, TIME_SCALES, TimeString, TimezoneInfo
+from .. import Time, ScaleValueError, TIME_SCALES, TIME_TYPES, TimeString, TimezoneInfo
```
As noted, I think `TIME_TYPES` should not be exported (since `TIME_DELTA_TYPES` is not either). Also, I think the tests are clearer if they do not use `TIME_TYPES['tt']` but rather `STANDARD_TIME_SCALES`. So,
```
from ..core import STANDARD_TIME_SCALES
```

mhvk commented on astropy/time/tests/test_basic.py:
```diff
@@ -317,10 +317,10 @@ def test_all_transforms(self):
         lat = 19.48125
         lon = -155.933222
-        for scale1 in TIME_SCALES:
+        for scale1 in TIME_TYPES['tt']:
```
use `STANDARD_TIME_SCALES` here and below

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -7,7 +7,7 @@
 
 from .. import (Time, TimeDelta, OperandTypeError, ScaleValueError,
-                TIME_SCALES, TIME_DELTA_SCALES)
+                TIME_SCALES,TIME_TYPES, TIME_DELTA_SCALES)
```
Same comment as above.

mhvk commented:
@vermashresth - now see your comment in #6487 - Above, I suggested importing `STANDARD_TIME_SCALES` and use that throughout. I still think that makes most sense, though I now see that it breaks some `table` tests as well as some documentation. For the latter, we probably should do
```
TIME_SCALES = STANDARD_TIME_SCALES + LOCAL_SCALES
```
and similarly for `TIME_DELTA_SCALES`. This will ensure the order is preserved, which is much easier for doctests.

For the table tests, the failures are related to a worry I expressed above: for formats with a reference time, it is assumed that the time passed in can be transformed to the scale of that reference time. For "local", this is not possible, and hence the format conversion fails. I think in principle, it may be possible to get those conversions to work, but as a first go, I think it is fine to edit the table test to work with `STANDARD_TIME_SCALES` and then add another test that uses a format that is safe for "local".

vermashresth commented:
Thanks for the review! @mhvk 
I will soon make the changes.

vermashresth commented:
Hi! @mhvk 
Sorry for the delay, had exams going on.
I have fixed the tests in `table` `io.fits` and doctests and now the build passes. Regarding the `io.fits` tests ([here](https://github.com/astropy/astropy/blob/master/astropy/io/fits/tests/test_fitstime.py#L341:L344)) it checks whether a warning is shown when `io.fits` `local` time scale is used but now `local` is supported natively so no warning will be shown. For now, I have removed the checking for warning part but should we get away completely with test for `local` in there?  
Will add tests for `local` next.

vermashresth commented:
Hi @mhvk  
I have added the tests. Build tests are also passing. Can you please review the changes?

mhvk commented on astropy/time/core.py:
```diff
@@ -1336,14 +1340,20 @@ def __sub__(self, other):
-                out._set_scale(other.scale if other.scale is not None
-                               else 'tai')
+                if other.scale is not None and self.scale not in TIME_TYPES[other.scale]:
+                    raise TypeError("Cannot subtract Time and TimeDelta instances with scales '{0}' and '{1}'".format(self.scale, other.scale))
```
This is a very long line. Could you break it up?

mhvk commented on astropy/time/core.py:
```diff
@@ -1336,14 +1340,20 @@ def __sub__(self, other):
         else:  # T - T
+
+            # the scales should be compatible (e.g., cannot convert TDB to LOCAL)
+            if(other.scale not in self.SCALES):
```
No reason for the parentheses, nor really for the black line (since there is a comment)

mhvk commented on astropy/time/core.py:
```diff
@@ -1386,8 +1396,11 @@ def __add__(self, other):
         else:
-            out._set_scale(other.scale if other.scale is not None else 'tai')
-
+            if other.scale is not None and self.scale not in TIME_TYPES[other.scale]:
```
Same comment on the flow; just start with `if other.scale is None` (as above).

mhvk commented on astropy/time/core.py:
```diff
@@ -1336,14 +1340,20 @@ def __sub__(self, other):
-                               else 'tai')
+                if other.scale is not None and self.scale not in TIME_TYPES[other.scale]:
+                    raise TypeError("Cannot subtract Time and TimeDelta instances with scales '{0}' and '{1}'".format(self.scale, other.scale))
+                else:
```
I think the logic is a bit odd here with an implicitly nested if.  I would do
```
if other.scale is None:
    out._set_scale('tai')
else:
    if self.scale not in TIME_TYPES[other.scale]:
        raise ...
    self._set_scale(other.scale)
```

mhvk commented on astropy/time/formats.py:
```diff
@@ -388,7 +388,14 @@ def to_value(self, parent=None):
+                tm = getattr(parent, self.epoch_scale)
+            except Exception as err:
+                raise ScaleValueError("Cannot convert from '{0}' epoch scale '{1}'"
+                                  "to specified scale '{2}', got error:\n{3}"
```
Please align properly.

mhvk commented on astropy/time/tests/test_basic.py:
```diff
@@ -350,6 +356,35 @@ def test_creating_all_formats(self):
+        ScalevalueError
+        """
+        t = Time('2006-01-15 21:24:37.5', scale='local')
+        assert t.jd == 2453751.3921006946
```
I worry this may (start to) fail due to rounding errors on some platforms.  Maybe
```
assert_allclose(t.jd, 2453751.3921006946, atol=0.001/3600./24., rtol=0.)
```

mhvk commented on astropy/time/tests/test_basic.py:
```diff
@@ -350,6 +356,35 @@ def test_creating_all_formats(self):
+        t = Time('2006-01-15 21:24:37.5', scale='local')
+        assert t.jd == 2453751.3921006946
+        assert t.mjd == 53750.892100694444
+        assert t.decimalyear == 2006.0408002758752
```
For the year, `atol` would need to be further divided by 365.

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -297,6 +298,40 @@ def test_scales_for_time_minus_time(self, scale1, scale2):
+        I.e., time differences of two times should have the scale of the
+        first time.
+
+        There are no local timescales for which this does not work.
```
This seems a silly statement, since there is only one!

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -297,6 +298,40 @@ def test_scales_for_time_minus_time(self, scale1, scale2):
+        assert t1_recover_t2_scale.scale == scale2
+        t1_recover = getattr(t1_recover_t2_scale, scale1)
+        assert allclose_jd(t1_recover.jd, t1.jd)
+        t2_recover_t1_scale = t1 - dt
```
Here, also add a test with a `TimeDelta(..., scale=None)` as well as just subtracting a quantity with time units.

mhvk commented on docs/time/index.rst:
```diff
@@ -264,7 +264,7 @@ both" [#]_. See also [#]_ and [#]_.
 
   >>> Time.SCALES
-  ('tai', 'tcb', 'tcg', 'tdb', 'tt', 'ut1', 'utc')
+  ('tai', 'tcb', 'tcg', 'tdb', 'tt', 'ut1', 'utc', 'local')
```
You should also adjust the text so that it becomes clear what 'local' is, maybe including the table below.

vermashresth commented:
Hi @mhvk 
I have made the changes you requested but the build is now failing and I can't figure out why. It shows some unrelated numpy related errors (and I noticed the same tests are failing in other PRs too). Can you please take a look?
Also, should we update the time scale conversion diagram in documentation?

mhvk commented on astropy/table/tests/test_mixin.py:
```diff
@@ -176,22 +176,85 @@ def test_io_time_write_fits(tmpdir, table_types):
+@pytest.mark.parametrize('table_types', (Table, QTable))
+def test_io_time_write_fits_local(tmpdir, table_types):
+    """
+    Test that table with Time mixin columns can be written by io.fits.
```
I think the docstring was just copy-pasted from the function above. Could you update it? (i.e., state that this tests the `local` time scale, and describe why it cannot be included above)?

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -268,15 +269,15 @@ def test_delta_scales_definition(self):
     @pytest.mark.parametrize(('scale1', 'scale2'),
-                             list(itertools.product(TIME_SCALES, TIME_SCALES)))
-    def test_scales_for_time_minus_time(self, scale1, scale2):
+                             list(itertools.product(STANDARD_TIME_SCALES, STANDARD_TIME_SCALES)))
```
nitpick: could you break the line? It is rather long now.

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -297,6 +298,45 @@ def test_scales_for_time_minus_time(self, scale1, scale2):
         assert allclose_jd(t2_recover.jd, t2.jd)
 
+    def test_local_scales_for_time_minus_time(self):
+        """T(X) - T2(Y)  -- does T(X) - T2(Y).X and return dT(X)
```
Please again update the docstring to reflect what *this* test does.

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -297,6 +298,45 @@ def test_scales_for_time_minus_time(self, scale1, scale2):
+        Also tests that subtracting two time scales, one of which is
+        different from local time scale should raise TypeError.
+        """
+        scale1 = 'local'
```
Rather than copy the test from above, can we just write 'local'? That makes the test easier to understand, and we're not iterating over different scales here.

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -297,6 +298,45 @@ def test_scales_for_time_minus_time(self, scale1, scale2):
+        # now check with delta time; also check reversibility
+        t1_recover_t2_scale = t2 + dt
+        assert t1_recover_t2_scale.scale == scale2
+        t1_recover = getattr(t1_recover_t2_scale, scale1)
```
This is a pointless operation when the scale is already 'local'; just call the above `t1_recover` instead of `t1_recover_t2_scale`.

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -297,6 +298,45 @@ def test_scales_for_time_minus_time(self, scale1, scale2):
+        assert allclose_jd(t1_recover.jd, t1.jd)
+        t2_recover_t1_scale = t1 - dt
+        assert t2_recover_t1_scale.scale == scale1
+        t2_recover = getattr(t2_recover_t1_scale, scale2)
```
This can be removed altogether, it is a pointless test.

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -297,6 +298,45 @@ def test_scales_for_time_minus_time(self, scale1, scale2):
+        assert t3.scale == t2.scale
+        q = 10 * u.s
+        assert (t2 - q).value == (t2 - dt2).value
+        for scale3 in STANDARD_TIME_SCALES:
```
This would become `for scale in ...`

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -359,6 +399,12 @@ def test_scales_for_delta_minus_delta(self):
                 dt_ut1 - self.dt[scale]
 
+        # local time scale
+        dt_local = self.dt['local']
```
Maybe add a test here that shows that `dt_local - dt_local` does work.

mhvk commented on docs/time/index.rst:
```diff
@@ -169,8 +169,8 @@ yday         :class:`~astropy.time.TimeYearDayTime`             2000:001:00:00:0
 .. note:: The :class:`~astropy.time.TimeFITS` format allows for most
-   but not all of the the FITS standard [#]_. Not implemented (yet) is
-   support for a ``LOCAL`` timescale. Furthermore, FITS supports some deprecated
+   but not all of the the FITS standard [#]_. It also supports the
```
Maybe 
```
.. note:: The :class:`~astropy.time.TimeFITS` format implements most
of the FITS standard [#]_, including support for the ``LOCAL`` timescale.
Note, though, that FITS supports some deprecated
```

mhvk commented on docs/time/index.rst:
```diff
@@ -276,13 +276,18 @@ tdb    Barycentric Dynamical Time  (TDB)
 tt     Terrestrial Time            (TT)
 ut1    Universal Time              (UT1)
 utc    Coordinated Universal Time  (UTC)
+local  Local Time Scale            (LOCAL)
```
Since this is typeset with the `(...)` part next to the text, maybe just right-adjust `(LOCAL)` to the end of the column (and move `(TT)` out by one space)

mhvk commented on docs/time/index.rst:
```diff
@@ -276,13 +276,18 @@ tdb    Barycentric Dynamical Time  (TDB)
+local  Local Time Scale            (LOCAL)
 ====== =================================
-
+  
```
Delete extraneous spaces.

mhvk commented on docs/time/index.rst:
```diff
@@ -276,13 +276,18 @@ tdb    Barycentric Dynamical Time  (TDB)
        `(PDF) <http://www.iausofa.org/sofa_ts_c.pdf>`_
 .. [#] `<http://www.ucolick.org/~sla/leapsecs/timescales.html>`_
 
+.. note:: ``local`` TimeScale is a scale that cannot be converted to/from any
```
Maybe nice to tell what it is meant for. E.g,
```
.. note:: the ``local`` time scale is meant for free-running clocks or simulation times,
  i.e., to represent a time without properly defined scale. It this cannot be converted
  to any other scale, and arithmetic is possible only with |Time| instances with scale
  ``local`` and with |TimeDelta| instances with scale ``local`` or `None`.
```
  
any other timescale)

mhvk commented on docs/time/index.rst:
```diff
@@ -276,13 +276,18 @@ tdb    Barycentric Dynamical Time  (TDB)
+   other scale and to which only TimeDelta with ``None`` type scale could be
+   added or subtracted.
+
 The system of transformation between supported time scales is shown in the
```
I think maybe here note again that ``local`` is excluded; then we do not have to update the figure. E.g.,
```
The system of transformations between the standard time scales (i.e., all but ``local``)
is shown in the ...
```

vermashresth commented:
Done :)
Time for entry into `CHANGES.rst`?

mhvk commented:
Indeed, time for a changelog entry!

vermashresth commented:
@mhvk Is there any more work needed on this?
And can you tell more details about using this in `io.fits` like you suggested. I would like to try implementing that too.

mhvk commented:
@vermashresth - I think it is all OK, but would like to have a last look. Unfortunately, haven't had time to do so yet. As for the comment about `io.fits`, you ended up doing that in the PR already: things will now properly translate to and from fits!

mhvk commented:
@vermashresth - I finally got around to a final review of your nice PR. To ensure we can merge it, I went ahead and just made a commit with some further small changes. I'll now merge, assuming tests pass.

mhvk commented:
OK, I guess I need to rebase as well so we can get the changelog properly set up...  Will do that.

mhvk commented:
OK, merged. Thanks, @vermashresth!

vermashresth commented:
Finally :D
Thanks!! @mhvk
